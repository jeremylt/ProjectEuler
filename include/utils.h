#ifndef UTILS_H
#define UTILS_H
#include <math.h>
#include <stdio.h>
#include <limits.h>
#include <stdlib.h>
#include <stdbool.h>

// -----------------------------------------------------------------------------
// Interface utilites
// -----------------------------------------------------------------------------
int getUserInt(char *userPrompt, FILE *stream, int *input);
int getValidUserInt(char *userPrompt, FILE *stream, int lower, int upper,
                    int *input);
int getUserLong(char *userPrompt, FILE *stream, long *input);
int getValidUserLong(char *userPrompt, FILE *stream, long lower, long upper,
                     long *input);

// -----------------------------------------------------------------------------
// Math utilities
// -----------------------------------------------------------------------------
long longMin(long a, long b);
long longMax(long a, long b);
long longPow(long base, long power);
int calculateGCD(long a, long b, long *result);
int calculateLCM(long a, long b, long *result);
int isPrime(long number, bool *prime);
int isPrimeFallback(long number, bool *prime);
int isPalindrome(long number, bool *palindrome);

#endif
