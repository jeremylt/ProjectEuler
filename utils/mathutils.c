#include "utils.h"
/// @file
/// Mathematical utilities
///
/// @addtogroup Math
/// @{
// -----------------------------------------------------------------------------
/**
  @brief Compute gcd of integers (long int) a and b

  @param[in] a        first integer
  @param[in] b        second integer
  @param[out] result  gcd of a and b

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int calculateGCD(long a, long b, long *result) {
  // Compute gcd
  if (b == 0)
    *result = a;
  else
    calculateGCD(b, a % b, result);

  return 0;
};
// -----------------------------------------------------------------------------
/**
  @brief Compute lcm of integers (long int) a and b

  @param[in] a        first integer
  @param[in] b        second integer
  @param[out] result  lcm of a and b

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int calculateLCM(long a, long b, long *result) {
  // Compute gcd
  long gcd;
  calculateGCD(a, b, &gcd);

  // Compute lcm
  long lcm = a * b / gcd;

  *result = lcm;

  return 0;
};
// -----------------------------------------------------------------------------
/**
  @brief Check if number (long int) is prime

  @param[in] number  number to check primality
  @param[out] prime  boolian value, true if prime, othewise false

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int isPrime(long number, bool *prime) {
  long bound = floor(sqrt((double)number));

  // Check if prime
  for (long i = bound; i > 1; i--) {
    *prime = number % i;
    if (!(*prime))
      break;
  }

  return 0;
};
// -----------------------------------------------------------------------------
/**
  @brief Check if number (long int) is a palindrome

  @param[in] number  number to check as palindrome
  @param[out] prime  boolian value, true if palindrome, othewise false

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int isPalindrome(long number, bool *palindrome) {
  int numdigits = floor(log10(number)) + 1;

  // Convert to array
  int *numasarray;
  numasarray = (int *)calloc(numdigits, sizeof(int));
  long copy = number;
  for (int i = 0; i < numdigits; i++) {
    numasarray[i] = copy % 10;
    copy /= 10;
  }

  // Check if palindrome
  for (int i = 0; i < numdigits / 2; i++) {
    *palindrome = (numasarray[i] == numasarray[numdigits-i-1]);
    if (!(*palindrome))
      break;
  }

  // Cleanup
  free(numasarray);

  return 0;
};
// -----------------------------------------------------------------------------
long longMin(long a, long b) {
  return a < b ? a : b;
};
/**
  @brief Return maximum of two integers (long int)

  @param[in] a  first integer
  @param[in] b  second integer

  @return  Maximum of a and b
**/
// -----------------------------------------------------------------------------
long longMax(long a, long b) {
  return a > b ? a : b;
};
// -----------------------------------------------------------------------------
/**
  @brief Return integer (long int) power

  @param[in] base   base to exponentiate
  @param[in] power  power to raise the base to

  @return base^power

  @ref Utility
**/
// -----------------------------------------------------------------------------
long longPow(long base, long power) {
  long result = 1;

  while (power) {
    if (power & 1) result *= base;
    power >>= 1;
    base *= base;
  }
  return result;
}
/// @}
