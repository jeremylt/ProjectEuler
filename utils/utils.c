#include "utils.h"
/// @file
/// Interface utilities
///
/// @addtogroup Utilities
/// @{
// -----------------------------------------------------------------------------
/**
  @brief Get user input integer

  @param[in] userPrompt  Text for user prompt
  @param[out] input      User input integer

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int getUserInt(char *userPrompt, FILE *stream, int *input) {
  char *p, s[100];
  int n = 0;

  // Get user input with given prompt
  printf("%s", userPrompt);
  while (fgets(s, sizeof(s), stream)) {
    n = strtol(s, &p, 10);
    // Check for integer
    if (p == s || *p != '\n') {
      printf("%s", userPrompt);
    } else break;
  }

  // Save user input
  *input = n;

  return 0;
};
// -----------------------------------------------------------------------------
/**
  @brief Get user input integer from a range

  @param[in] userPrompt  Text for user prompt
  @param[in] lower       Lower end of range, inclusive
  @param[in] upper       Upper end of range, inclusive
  @param[out] input      User input integer

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int getValidUserInt(char *userPrompt, FILE *stream, int lower, int upper,
                    int *input) {
  bool valid = false;

  // Loop until valid
  while (!valid) {
    getUserInt(userPrompt, stream, input);
    if (*input < lower || *input > upper)
      printf("Invalid input! Provide a value in the range [%d - %d].\n", lower,
             upper);
    else
      break;
  }

  return 0;
};
// -----------------------------------------------------------------------------
/**
  @brief Get user input integer (long int)

  @param[in] userPrompt  Text for user prompt
  @param[out] input      User input integer

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int getUserLong(char *userPrompt, FILE *stream, long *input) {
  char *p, s[100];
  long n = 0;

  // Get user input with given prompt
  printf("%s", userPrompt);
  while (fgets(s, sizeof(s), stream)) {
    n = strtol(s, &p, 10);
    // Check for integer
    if (p == s || *p != '\n') {
      printf("%s", userPrompt);
    } else break;
  }

  // Save user input
  *input = n;

  return 0;
};
// -----------------------------------------------------------------------------
/**
  @brief Get user input integer (long int) from a range

  @param[in] userPrompt  Text for user prompt
  @param[in] lower       Lower end of range, inclusive
  @param[in] upper       Upper end of range, inclusive
  @param[out] input      User input integer

  @return  An error code: 0 - success, otherwise - failure
**/
// -----------------------------------------------------------------------------
int getValidUserLong(char *userPrompt, FILE *stream, long lower, long upper,
                     long *input) {
  bool valid = false;

  // Loop until valid
  while (!valid) {
    getUserLong(userPrompt, stream, input);
    if (*input < lower || *input > upper)
      printf("Invalid input! Provide a value in the range [%ld - %ld].\n", lower,
             upper);
    else
      break;
  }

  return 0;
};
// -----------------------------------------------------------------------------
/// @}
