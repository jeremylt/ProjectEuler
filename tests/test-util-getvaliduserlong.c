/// @file
/// Test Project Euler getValidUserLong
/// \test Test getValidUserLong result
#include <stdio.h>
#include "utils.h"
// -----------------------------------------------------------------------------
int main(void) {
  long trueResult = 7;
  FILE *stream;
  long lower = 1, upper = 500, input;

  // Open file
  stream = fopen("tests/input/test-util-getvaliduserlong.in" , "r");
  if(stream == NULL) {
    perror("Error opening file\n");
    return(-1);
  }

  // Get userInt
  getValidUserLong("", stream, lower, upper, &input);

  // Check
  if (input != trueResult)
    printf("Error:\n\tComputed: %ld != True: %ld\n", input, trueResult);

  // Close
  fclose(stream);

  return 0;
}
// -----------------------------------------------------------------------------
