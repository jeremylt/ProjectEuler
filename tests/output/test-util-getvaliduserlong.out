Invalid input! Provide a value in the range [1 - 500].
